pennylane/grouping/pauli_group.py

```diff
@@ -0,0 +1,273 @@
+    Pauli words, rather than matrices. For example,
+
+    .. code-block:: python
+
+        import pennylane as qml
+        from pennylane.grouping.pauli_group import pauli_mult
+
+        pauli_1 = qml.PauliX(0) @ qml.PauliZ(1)
+        pauli_2 = qml.PauliY(0) @ qml.PauliZ(1)
+
+        product = pauli_mult(pauli_1, pauli_2)
+        print(product)
+
+    will yield ``qml.PauliZ(0)``.
+    """
+
+    # If no wire map is specified, generate one from the union of wires
+    # in both Paulis.
+    if wire_map is None:
+        wire_labels = set(pauli_1.wires.labels + pauli_2.wires.labels)
+        wire_map = {label: i for i, label in enumerate(wire_labels)}
+
+    # Check if pauli_1 and pauli_2 are the same; if so, the result is the Identity
+    if are_identical_pauli_words(pauli_1, pauli_2):
+        first_wire = list(wire_map.keys())[0]
+        return Identity(first_wire)
+
+    # Compute binary symplectic representations
+    pauli_1_binary = pauli_to_binary(pauli_1, wire_map=wire_map)
+    pauli_2_binary = pauli_to_binary(pauli_2, wire_map=wire_map)
+
+    bin_symp_1 = np.array([int(x) for x in pauli_1_binary])
+    bin_symp_2 = np.array([int(x) for x in pauli_2_binary])
+
+    # Shorthand for bitwise XOR of numpy arrays
+    pauli_product = bin_symp_1 ^ bin_symp_2
+
+    return binary_to_pauli(pauli_product, wire_map=wire_map)
+
+
+def pauli_mult_with_phase(pauli_1, pauli_2, wire_map=None):
+    r"""Multiply two Pauli words together including the global phase.
+
+    Two Pauli operations can be multiplied together by taking the additive
+    OR of their binary symplectic representations. The phase is computed by
+    looking at the number of times we have the products  XY, YZ, or ZX (adds a
+    phase of :math:`i`), or YX, ZY, XZ (adds a phase of :math:`-i`).
+
+    Args:
+        pauli_1 (qml.Operation): A Pauli word.
+        pauli_2 (qml.Operation): A Pauli word to multiply with the first one.
+        wire_map  (dict[Union[str, int], int]): dictionary containing all wire labels used in the Pauli
+            word as keys, and unique integer labels as their values. If no wire map is
+            provided, the map will be constructed from the set of wires acted on
+            by the input Pauli words.
+
+    Returns:
+        (qml.Operation, np.complex): The product of pauli_1 and pauli_2, and the
+            global phase.
+
+    **Example**
+
+    This function works the same as ``pauli_mult`` but also returns the global
+    phase accumulated as a result of the Pauli product rules
+    :math:`\sigma_i \sigma_j = i \sigma_k`.
+
+    .. code-block:: python
+
+        import pennylane as qml
+        from pennylane.grouping.pauli_group import pauli_mult
+
+        pauli_1 = qml.PauliX(0) @ qml.PauliZ(1)
+        pauli_2 = qml.PauliY(0) @ qml.PauliZ(1)
+
+        product, phase = pauli_mult_with_phase(pauli_1, pauli_2)
+
+    will yield ``qml.PauliZ(0)`` and :math:`1j`.
```

Comment:
Where it makes sense, it's always preferable to use pycon syntax for docstring examples (this follows the convention the NumPy docs use)
```suggestion
    >>> from pennylane.grouping.pauli_group import pauli_mult_with_phase
    >>> pauli_1 = qml.PauliX(0) @ qml.PauliZ(1)
    >>> pauli_2 = qml.PauliY(0) @ qml.PauliZ(1)
    >>> product, phase = pauli_mult_with_phase(pauli_1, pauli_2)
    >>> product
    PauliZ(wires=[0])
    >>> phase
    1j
```